designs/provisioner-priority.md

```diff
@@ -0,0 +1,196 @@
+
+   _Note: It is still possible that pods with tolerations may not be scheduled to nodes with taints even if they are preferred, since pods with tolerations can technically be scheduled anywhere they are tolerated (including nodes that contain no taints)._
+
+   **Example**
+
+    ```yaml
+    # Custom Workload for Pods that Tolerate these Taints should be attempted first
+
+    apiVersion: karpenter.sh/v1alpha5
+    kind: Provisioner
+    metadata:
+      name: 
+    spec:
+      weight: 100
+      requirements:
+      - key: node.kubernetes.io/instance-type
+        operator: In
+        values: ["p3.8xlarge", "p3.16xlarge"]
+      taints:
+      - key: custom-workloads
+        value: "true"
+        effect: NoSchedule
+    ```
+
+    ```yaml
+    # Workloads that do not have any specific requirements can pick up this provisioner as the backup
+
+    apiVersion: karpenter.sh/v1alpha5
+    kind: Provisioner
+    metadata:
+      name: fallback
+    spec:
+      weight: 50
+      requirements:
+      - key: kubernetes.io/arch
+        operator: In
+        values: ["amd64"]
+    ```
+
+## Proposed Design
+
+To enable the ability to define a user-defined relationship between provisioners that will be considered in scheduling, we introduce a `.spec.weight` value in the `karpenter.sh/v1alpha5/Karpenter` provisioner spec. This value will have the following constraints:
+
+1. The provisioner weight value will be an integer from 1-100 if specified
+2. A provisioner with no weight is considered to be a provisioner with weight 0
+3. Provisioners with the same weight have no guarantee on ordering and will be randomly ordered
+
+These constraints are consistent with pod affinity/pod anti-affinity preference weights described [here](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#an-example-of-a-pod-that-uses-pod-affinity).
+
+## Considerations
+
+### Provisioner Scheduling Ordering
+
+__Current State__: Scheduling calls the Kubernetes LIST API for the `karpenter.sh/v1alpha5/Provisioner` and receives a random ordering of Provisioners back for scheduling. When processing pods, Karpenter will find the first provisioner that it is able to schedule the pod onto given the instance type options that are constrained by the pod scheduling requirements (affinity, anti-affinity, topology spread) and the provisioner requirements. This creates some randomness for users and leads to some unpredictability with respect to the provisioner that will be used for a given pod. Currently, the best practices that are recommended by the [Karpenter/EKS docs](https://aws.github.io/aws-eks-best-practices/karpenter/#create-provisioners-that-are-mutually-exclusive) recommend that we create mutually exclusive provisioners such that we reduce this unpredictability.
+
+**Improvement Options**
```

Comment:
You might add an option where `Provisioner` matching is scored for an implicit weight rather than the simple `Compatible` check we do today. This would solve for preferring more requirements matches like in the taints case above.

Reply:
I'm a bit concerned about how this will affect the bin-packing logic. In general, since we are bin-packing on nodes we have already decided to provision, if we get new pods, we are going to add to the current node, not prefer other provisioners over this current one